snakemake/script.py

```diff
@@ -381,47 +410,127 @@ def generate_preamble(
         shadow_dir,
         preamble_addendum="",
     ):
-        wrapper_path = path[7:] if path.startswith("file://") else path
-        snakemake = Snakemake(
-            input_,
-            output,
-            params,
-            wildcards,
-            threads,
-            resources,
-            log,
-            config,
-            rulename,
-            bench_iteration,
-            os.path.dirname(wrapper_path),
-        )
-        snakemake = pickle.dumps(snakemake)
-        # Obtain search path for current snakemake module.
-        # The module is needed for unpickling in the script.
-        # We append it at the end (as a fallback).
+
+        # Obtain search path for any mounted containers
+        # This will also ensure recipes that require snakemake have it on path
         searchpath = SNAKEMAKE_SEARCHPATH
         if container_img is not None:
             searchpath = singularity.SNAKEMAKE_MOUNTPOINT
         searchpath = repr(searchpath)
+
         # For local scripts, add their location to the path in case they use path-based imports
-        if path.startswith("file://"):
+        if path.startswith(file_prefix):
             searchpath += ", " + repr(os.path.dirname(path[7:]))
 
+        # The wrapper path should be added too
+        wrapper_path = path[7:] if path.startswith(file_prefix) else path
+
         return textwrap.dedent(
             """
         ######## snakemake preamble start (automatically inserted, do not edit) ########
-        import sys; sys.path.extend([{searchpath}]); import pickle; snakemake = pickle.loads({snakemake}); from snakemake.logging import logger; logger.printshellcmds = {printshellcmds}; {preamble_addendum}
+        import sys, os; sys.path.extend([{searchpath}]); sys.path.insert(0, {sourcedir});
+        class Snakemake:
```

Comment:
The entire template should be put into a separate file now, also the R and Julia ones. Otherwise, this just becomes too big.

Reply:
Should we turn script into a folder module, akin to executors?